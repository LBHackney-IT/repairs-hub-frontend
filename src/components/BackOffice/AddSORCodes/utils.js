import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

export const fetchContractors = () =>
  new Promise(async (resolve) => {
    const contractors = await frontEndApiRequest({
      method: 'get',
      path: '/api/contractors?getAllContractors=true',
    })

    resolve(contractors)
  })

export const fetchTrades = () =>
  new Promise(async (resolve) => {
    const trades = await frontEndApiRequest({
      method: 'get',
      path: '/api/schedule-of-rates/trades?getAllTrades=true',
    })

    resolve(trades)
  })

export const fetchContracts = (contractorReference) =>
  new Promise(async (resolve) => {
    const contracts = await frontEndApiRequest({
      method: 'get',
      path: `/api/backoffice/contracts?contractorReference=${contractorReference}`,
    })

    resolve(contracts)
  })

export const saveSorCodesToDatabase = async (data) => {
  await frontEndApiRequest({
    method: 'post',
    path: `/api/backoffice/sor-codes`,
    requestData: data,
  })
}

export const dataToRequestObject = (csvArray, contractReference, tradeCode) => {
  const sorCodes = csvArray.map((x) => ({
    code: x.Code,
    cost: parseFloat(x.Cost),
    standardMinuteValue: parseFloat(x.StandardMinuteValue),
    shortDescription: x.ShortDescription,
    longDescription: x.LongDescription,
  }))

  return {
    contractReference: contractReference,
    tradeCode: tradeCode,
    sorCodes: sorCodes,
    contract: 'placeholder',
  }
}
