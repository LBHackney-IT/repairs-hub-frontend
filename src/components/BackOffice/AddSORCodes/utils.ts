import { frontEndApiRequest } from '@/root/src/utils/frontEndApiClient/requests'

import { Contractor, Trade, NewSorCodesRequestObject } from './types'

export const fetchContractors = async () : Promise<Array<Contractor>>  => {
  return await frontEndApiRequest({
    method: 'get',
    path: '/api/contractors?getAllContractors=true',
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const fetchTrades = async () : Promise<Array<Trade>> => {
  return await frontEndApiRequest({
    method: 'get',
    path: '/api/schedule-of-rates/trades?getAllTrades=true',
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const fetchContracts = async (contractorReference: string) : Promise<Array<string>> => {
  return await frontEndApiRequest({
    method: 'get',
    path: `/api/backoffice/contracts?contractorReference=${contractorReference}`,
    params: null,
    paramsSerializer: null,
    requestData: null
  })
}

export const saveSorCodesToDatabase = async (data: NewSorCodesRequestObject): Promise<void> => {
  await frontEndApiRequest({
    method: 'post',
    path: `/api/backoffice/sor-codes`,
    requestData: data,
    params: null,
    paramsSerializer: null,
  })
}

export const dataToRequestObject = (csvArray: Array<any>, contractReference: string, tradeCode: string): NewSorCodesRequestObject => {
  const sorCodes = csvArray.map((x) => ({
    code: x.Code,
    cost: parseFloat(x.Cost),
    standardMinuteValue: parseFloat(x.StandardMinuteValue),
    shortDescription: x.ShortDescription,
    longDescription: x.LongDescription,
  }))

  return {
    contractReference: contractReference,
    tradeCode: tradeCode,
    sorCodes: sorCodes,
  }
}
