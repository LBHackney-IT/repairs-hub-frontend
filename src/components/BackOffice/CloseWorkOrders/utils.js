export const formatWorkOrderReferences = (workOrderReferences) => {
  return [
    ...new Set(
      workOrderReferences
        .split('\n')
        .map((x) => x.trim())
        .filter((x) => x)
    ),
  ]
}

const validateWorkOrderReference = (workOrderReference) => {
  const regex = /^\d{8}$/gm

  return regex.exec(workOrderReference) !== null
}

export const getInvalidWorkOrderReferences = (workOrderReferences) => {
  return workOrderReferences
    .filter((x) => !validateWorkOrderReference(x))
    .map((x) => `"${x}"`)
}

export const formatInvalidWorkOrderReferencesError = (
  invalidWorkOrderReferences
) => {
  return `Invalid WorkOrder Reference(s) entered: ${invalidWorkOrderReferences.join(
    ', '
  )}`
}

export const dateIsInFuture = (date) => {
  const today = new Date()

  return date > today
}
