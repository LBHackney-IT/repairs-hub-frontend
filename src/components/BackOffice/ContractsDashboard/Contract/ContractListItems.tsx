import Contract from '@/root/src/models/contract'

import WarningInfoBox from '../../../Template/WarningInfoBox'
import ContractListItem from './ContractListItem'
import ErrorMessage from '../../../Errors/ErrorMessage'

interface Props {
  contracts: Contract[]
  warningText?: string
  error?: Error | string | null
  page: string
  activeStatus?: string
}

const ContractListItems = (props: Props) => {
  const { contracts, warningText, error, page, activeStatus } = props

  return (
    <>
      {(contracts === null || contracts?.length === 0) && (
        <div style={{ width: '90%' }}>
          <WarningInfoBox
            header="No contracts found!"
            text={`${warningText}`}
            name="no-contracts-found"
          />
        </div>
      )}

      {contracts?.length > 0 && (
        <ol
          style={{
            display: 'grid',
            gridTemplateColumns: 'repeat(auto-fill, 10rem)',
            gap: '1rem',
            listStyle: 'none',
            padding: 0,
            margin: 0,
          }}
          data-test-id={
            activeStatus ? `${activeStatus}-contracts-list` : 'contract-list'
          }
        >
          {contracts?.map((contract, index) => (
            <ContractListItem
              key={contract.contractReference}
              contract={contract}
              index={index}
              page={page}
            />
          ))}
        </ol>
      )}

      {error && (
        <ErrorMessage
          label={
            error instanceof Error
              ? error.message
              : typeof error === 'string'
              ? error
              : 'An unexpected error occurred'
          }
        />
      )}
    </>
  )
}

export default ContractListItems
