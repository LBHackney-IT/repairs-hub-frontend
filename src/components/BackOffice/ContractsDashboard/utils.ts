import Contract from '@/root/src/models/contract'

export const today = new Date()

export function monthsOffset(months: number) {
  return new Date(
    today.getFullYear(),
    today.getMonth() + months,
    today.getDate()
  )
}

export const filterContractsByExpiryDate = (
  contracts: Contract[],
  months: number
) => {
  const contractExpiryCutOffDate = monthsOffset(months)

  const filteredContracts = contracts?.filter((contract) => {
    return (
      new Date(contract.terminationDate) > today &&
      new Date(contract.terminationDate) < contractExpiryCutOffDate
    )
  })
  return filteredContracts
}

export const filterRelativeInactiveContracts = (
  contracts: Contract[],
  yearAfter: string
) => {
  const relativeInactiveContracts = contracts?.filter(
    (contract) =>
      contract.terminationDate > yearAfter &&
      contract.terminationDate < new Date().toISOString()
  )
  return relativeInactiveContracts
}
