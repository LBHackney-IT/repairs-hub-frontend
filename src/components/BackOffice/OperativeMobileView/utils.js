export const filterOperatives = (operatives, operativeFilter) => {
  if (!operativeFilter) return operatives;

  return operatives.filter((x) =>
    x.id?.toString().includes(operativeFilter) ||
    x.name?.toLowerCase().includes(operativeFilter) ||
    x.payrollNumber?.toLowerCase().includes(operativeFilter)
  );
}
