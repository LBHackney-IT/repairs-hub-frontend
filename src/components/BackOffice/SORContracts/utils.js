export const sanitizeInput = value => {
    return value.trim().replaceAll(' ', '')
}

export const saveContractChangesToDatabase = async (data) => {
    await frontEndApiRequest({
        method: 'POST',
        path: `/api/backOffice/sor-contracts`,
        requestData: data,
      })
  }