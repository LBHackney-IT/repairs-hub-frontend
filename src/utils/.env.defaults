### Service APIs
# Should be set in .env.local
REPAIRS_SERVICE_API_URL=http://localhost:5101/api
REPAIRS_SERVICE_API_KEY=
CONFIGURATION_API_URL=http://localhost:5102/api

### Single Sign On
# Should be set in .env
GSSO_URL=
HACKNEY_JWT_SECRET=
GSSO_TOKEN_NAME=token

### Application environment
NEXT_PUBLIC_ENV_NAME=localdev

### Node environment
NEXT_PUBLIC_NODE_ENV=development
NODE_ENV=development

### Redirect URL to run code locally
REDIRECT_URL=localdev.hackney.gov.uk:5000

NEXT_PUBLIC_MMH_FRONTEND_URL=https://manage-my-home-development.hackney.gov.uk/
CONTACT_DETAILS_API_URL=https://gos4l9my1a.execute-api.eu-west-2.amazonaws.com/development

### Feature flags
NEXT_PUBLIC_USE_DEPRECATED_PROPERTY_SEARCH=false
NEXT_PUBLIC_BUDGET_CODE_SELECTION_ENABLED=true

### Feature page sizing
NEXT_PUBLIC_PROPERTIES_PAGE_SIZE=50
CYPRESS_NEXT_PUBLIC_PROPERTIES_PAGE_SIZE=50
PAGE_SIZE_CONTRACTORS=10
CYPRESS_PAGE_SIZE_CONTRACTORS=10
PAGE_SIZE_AGENTS=10
CYPRESS_PAGE_SIZE_AGENTS=10


# ### Cypress auth tokens
# [You will need to populate these credentials in .env.local for cypress tests]

# #CYPRESS_GSSO_TEST_KEY_AGENT=
# #CYPRESS_GSSO_TEST_KEY_CONTRACTOR_ALPHATRACK=
# #CYPRESS_GSSO_TEST_KEY_MULTIPLE_CONTRACTOR=
# #CYPRESS_GSSO_TEST_KEY_CONTRACT_MANAGER=
# #CYPRESS_GSSO_TEST_KEY_AUTHORISATION_MANAGER=
# #CYPRESS_GSSO_TEST_KEY_DLO_CONTRACTOR=
# #CYPRESS_GSSO_TEST_KEY_OPERATIVE=
# #CYPRESS_GSSO_TEST_KEY_DATA_ADMIN=

### Feature configuration
NEXT_PUBLIC_STATIC_IMAGES_BUCKET_URL=
NEXT_PUBLIC_OUT_OF_HOURS_LINK=https://hackney.gov.uk
CYPRESS_NEXT_PUBLIC_OUT_OF_HOURS_LINK=https://hackney.gov.uk
NEXT_PUBLIC_SUPPORT_LINK=https://sites.google.com/hackney.gov.uk/repairshubsupport
NEXT_PUBLIC_DRS_SESSION_COOKIE_NAME=drs_session_id
CYPRESS_NEXT_PUBLIC_DRS_SESSION_COOKIE_NAME=drs_session_id