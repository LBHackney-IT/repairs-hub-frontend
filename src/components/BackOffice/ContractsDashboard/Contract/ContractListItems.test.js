import { render } from '@testing-library/react'

import { filterContractsByExpiryDate } from '../utils'

import ContractListItems from './ContractListItems'

import {
  mockActiveContracts,
  mockInactiveContracts,
} from '../mockContractsData'

describe('Contractors list items component', () => {
  it('should render active contracts with relevant fields', async () => {
    const { asFragment } = render(
      <ContractListItems
        contracts={mockActiveContracts}
        heading="Active Contracts:"
        warningText="No active contracts found for Syracuse Limited"
        page="contractor"
        activeStatus="active"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render inactive contracts with relevant fields', async () => {
    const { asFragment } = render(
      <ContractListItems
        contracts={mockInactiveContracts}
        heading="Inactive Contracts:"
        warningText="No inactive contracts found for Syracuse Limited"
        page="contractor"
        activeStatus="active"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render contracts that expire in two months with relevant fields', async () => {
    const contractsThatExpireInTwoMonths = filterContractsByExpiryDate(
      mockActiveContracts,
      2,
      new Date('2025-07-09T15:38:48.061Z')
    )
    const { asFragment } = render(
      <ContractListItems
        contracts={contractsThatExpireInTwoMonths}
        heading="Contracts due to expire soon:"
        warningText="No contracts expiring in the next two months."
        page="dashboard"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
