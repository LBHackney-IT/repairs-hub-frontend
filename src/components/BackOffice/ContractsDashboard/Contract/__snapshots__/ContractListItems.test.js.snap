// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contractors list items component should render active contracts with relevant fields 1`] = `
<DocumentFragment>
  <h3
    class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-bottom-1"
  >
    Active Contracts:
  </h3>
  <ol
    data-test-id="active-contracts-list"
    style="display: grid; grid-template-columns: repeat(auto-fill, 10rem); gap: 1rem; list-style: none; padding: 0px; margin: 0px;"
  >
    <li
      data-id="0"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          127-127-1277
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £0.00
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           0
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
    <li
      data-id="1"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          128-128-1288
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           34
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
    <li
      data-id="2"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          129-129-1299
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £544,898.23
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           3441
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
    <li
      data-id="3"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          130-130-1300
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £573,623.97
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           3443
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
    <li
      data-id="4"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          131-131-1311
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £143,393.57
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           904
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
    <li
      data-id="5"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          132-132-1322
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £56.74
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           2
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expires: 
          </span>
          9 Jan 2026
        </p>
      </div>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`Contractors list items component should render contracts that expire in two months with relevant fields 1`] = `
<DocumentFragment>
  <h3
    class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-bottom-1"
  >
    Contracts due to expire soon:
  </h3>
  <div
    style="width: 90%;"
  >
    <div
      class="warning-info-box govuk-inset-text lbh-inset-text"
      data-testid="no-contracts-found"
    >
      <div
        class="lbh-warning-text govuk-warning-text"
      >
        <span
          aria-hidden="true"
          class="govuk-warning-text__icon"
        >
          !
        </span>
        <div
          class="govuk-warning-text__text"
        >
          <span
            class="govuk-warning-text__assistive"
          >
            Warning
          </span>
          <p
            class="govuk-!-margin-top-0 lbh-body-s lbh-!-font-weight-bold"
          >
            No contracts found!
          </p>
          <p
            class="lbh-body-xs govuk-!-margin-top-1"
          >
            No contracts expiring in the next two months.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Contractors list items component should render inactive contracts with relevant fields 1`] = `
<DocumentFragment>
  <h3
    class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-bottom-1"
  >
    Inactive Contracts:
  </h3>
  <ol
    data-test-id="active-contracts-list"
    style="display: grid; grid-template-columns: repeat(auto-fill, 10rem); gap: 1rem; list-style: none; padding: 0px; margin: 0px;"
  >
    <li
      data-id="0"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          111-111-1111
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           20
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expired: 
          </span>
          31 Mar 2015
        </p>
      </div>
    </li>
    <li
      data-id="1"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          122-122-1222
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           20
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expired: 
          </span>
          31 Mar 2015
        </p>
      </div>
    </li>
    <li
      data-id="2"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          133-133-1388
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           20
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expired: 
          </span>
          9 May 2025
        </p>
      </div>
    </li>
    <li
      data-id="3"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          144-144-1444
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           20
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expired: 
          </span>
          9 May 2025
        </p>
      </div>
    </li>
    <li
      data-id="4"
      style="border: 5px solid #00664f; border-radius: 20px; box-sizing: border-box; padding: 1rem; display: flex; margin-top: 1rem; justify-content: center;"
    >
      <div
        class="contract-details"
      >
        <h5
          style="white-space: nowrap;"
        >
          155-155-1555
        </h5>
        <p>
          <span
            style="font-weight: 600;"
          >
            Sum of SORs:
          </span>
           £5,958.38
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            SOR Count:
          </span>
           20
        </p>
        <p>
          <span
            style="font-weight: 600;"
          >
            Expired: 
          </span>
          9 May 2025
        </p>
      </div>
    </li>
  </ol>
</DocumentFragment>
`;
